src/Products/Viewer/Controllers/ViewerApiController.cs

```diff
@@ -383,196 +411,318 @@ public HttpResponseMessage PrintPdf(PostedDataEntity loadDocumentRequest)
+                {
+                    MemoryPageStreamFactory pageStreamFactory = new MemoryPageStreamFactory(pages);
+
+                    if (globalConfiguration.Viewer.GetIsHtmlMode())
+                    {
+                        ViewOptions viewOptions = HtmlViewOptions.ForEmbeddedResources(pageStreamFactory);
+
+                        viewer.View(viewOptions);
+                    }
+                    else
+                    {
+                        PngViewOptions pngViewOptions = new PngViewOptions(pageStreamFactory);
+
+                        viewer.View(pngViewOptions);
+                    }
+                }
+
+                foreach (var pageStream in pages)
+                {
+                    if (globalConfiguration.Viewer.GetIsHtmlMode())
+                    {
+                        pagesContent.Add(Encoding.UTF8.GetString(pageStream.ToArray()));
+                    }
+                    else
+                    {
+                        pagesContent.Add(Convert.ToBase64String(pageStream.ToArray()));
+                    }
+                }
             }
-            foreach (PageData page in documentInfoContainer.Pages)
+
+            foreach (Page page in viewInfo.Pages)
             {
-                PageDescriptionEntity pageData = GetPageDescriptionEntities(page);
+                PageDescriptionEntity pageData = GetPageDescriptionEntities(page, pagesInfoPath);
                 if (pagesContent.Count > 0)
                 {
                     pageData.SetData(pagesContent[page.Number - 1]);
                 }
                 loadDocumentEntity.SetPages(pageData);
             }
+
             loadDocumentEntity.SetGuid(documentGuid);
-            loadDocumentEntity.SetShowGridLines(globalConfiguration.Viewer.GetShowGridLines());
-            // return document description
+
             return loadDocumentEntity;
         }
 
-
-        private dynamic GetHandler()
+        internal class MemoryPageStreamFactory : IPageStreamFactory
         {
-            if (viewerHtmlHandler != null)
+            private readonly List<MemoryStream> _pages;
+
+            public MemoryPageStreamFactory(List<MemoryStream> pages)
+            {
+                _pages = pages;
+            }
+
+            public Stream CreatePageStream(int pageNumber)
             {
-                return viewerHtmlHandler;
+                MemoryStream pageStream = new MemoryStream();
+
+                _pages.Add(pageStream);
+
+                return pageStream;
             }
-            else
+
+            public void ReleasePageStream(int pageNumber, Stream pageStream)
             {
-                return viewerImageHandler;
+                //Do not release page stream as we'll need to keep the stream open
             }
         }
 
-        private PageDescriptionEntity GetPageDescriptionEntities(PageData page)
+        private PageDescriptionEntity GetPageDescriptionEntities(Page page, string pagesInfoPath)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Make 'GetPageDescriptionEntities' a static method.](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5242614)